### ISSUE TITLE
Generalise beyond (U.Vector Double) for particle types

### ISSUE THREAD

**Issue by eggplantbren** [eggplantbren/NestedSampling.hs#12]
There are problems where the hypothesis space is not R^n, so the fact a particle is (U.Vector Double) shouldn't be hard coded. Any problem where you can define fromPrior, perturb, and logLikelihood should be usable. My gut still says we need a `class` for this but I could be wrong.

**Comment by jtobin** [comment]
The most general way to do this is to get rid of the `Particle` type and just replace it everywhere by a type parameter `a`.  So you wind up getting e.g.:

```
type Particles a = IntPSQ Lltb a
type Perturber a = a -> Gen RealWorld -> IO (Double, a)

data Sampler a = Sampler {
    samplerDim        :: {-# UNPACK #-} !Int
  , samplerSteps      :: {-# UNPACK #-} !Int
  , samplerLikelihood :: a -> Double
  , samplerPerturber  :: Perturber a
  , samplerParticles  :: !(Particles a)
  , samplerIter       :: {-# UNPACK #-} !Int
  , samplerLogZ       :: {-# UNPACK #-} !Double
  , samplerInfo       :: {-# UNPACK #-} !Double
  }
```

If you replace everything consistently like that, voilà.  Super general.  It will work over literally anything you can code up fromPrior/logLikelihood/perturb for.

A limitation is that, since we're making the domain so general (there are *no* restrictions on what it can be), we don't automatically know how to display points in it offhand.  This only becomes an issue in the following part of `writeToFile`, where we log the parameters to disk:

```
case logParametersFile of
  Nothing   -> return ()
  Just file -> do
    sample <- openFile file mode
    hPutStrLn sample $
      -- the following won't work because we don't know
      -- that 'particle' is going to be an unboxed vector
      drop 1 $ U.foldl' (\str x -> str ++ "," ++ show x) [] particle
    hClose sample
```

You can get around this in a couple of ways: don't log the parameters at all, or introduce constraints as to what the type of `a` can be.  For example, that it be an instance of `Show` or similar.

Either is acceptable, depending on how general you want the implementation to be.  We'll still get the standard output in e.g. `nested_sampling_info.csv`, so if you want extreme generality we could just log that and forget about `nested_sampling_parameters.csv`.  If that's too general and it'd be nice to have a look at our samples, we can implement the required constraints there.

**Comment by jtobin** [comment]
Resolved in #14.
